app.py

```diff
@@ -1039,6 +1039,69 @@ def timeout_handler(signum, frame):
 
         return cost, path, total_time
 
+    def nearest_neighbor(graph, order):
+        """
+        Implements the Nearest Neightbor Heuristic for TSP.
+        """
+        final_path = None
+        final_cost = INFINITY
+
+        # create a path for every single starting node
+        for key in graph.keys():
+            queue = []
+            list = order.copy()
+            first_node = (key[0], key[2])
+            queue.append(first_node)
+            total_cost = 0;
+
+            while list:
+                popped_node = queue[-1:]
+
+                # first time through, set the starting node as unvisited, used for cycling
+                if (total_cost == 0):
+                    queue.pop()
+                min_cost = INFINITY
+                visited_min_cost = INFINITY
+                next_node = None
+                visited_next_node = None
+
+                for (curr_node, dest_node, curr_dir), values in graph.items():
+
+                    # there exists an unvisited node, prioritize it
+                    if ( popped_node[0][0] == curr_node and popped_node[0][1] == curr_dir and (dest_node in list) ):
+                        for direc in values:
+                            if (min_cost > values[direc]['cost']):
+                                min_cost = values[direc]['cost']
+                                next_node = (dest_node, direc)
+
+                    # all nodes are visited, choose the least cost of the visited nodes
+                    elif ( popped_node[0][0] == curr_node and popped_node[0][1] == curr_dir ):
+                        for direc in values:
+                            if (visited_min_cost > values[direc]['cost']):
+                                visited_min_cost = values[direc]['cost']
+                                visited_next_node = (dest_node, direc)
+
+                # is there exists a path to a unvisited node, prioritize by adding it
+                # else, add the least cost path to a visited node
+                if (next_node is not None):
+                    total_cost += min_cost
+                    queue.append(next_node)
+                    if next_node[0] in list:
+                        list.remove(next_node[0])
+                else:
+                    total_cost += visited_min_cost
+                    queue.append(visited_next_node)
+                    if visited_next_node[0] in list:
+                        list.remove(visited_next_node[0])

```

Comment:
What is this trying to do? `if visited_next_node[0] in list:` isn't valid python logic.